### PR TITLE
add ucsc_blat tool

### PR DESCRIPTION
**Pull request by Yating-L**
Galaxy wrapper for UCSC BLAT alignment tool
===========================================
BLAT (http://genome.ucsc.edu/goldenPath/help/blatSpec.html) is a bioinformatics software a tool which performs rapid mRNA/DNA and cross-species protein alignments. 

Features
--------
1. Perform rapid mRNA/DNA and cross-species protein alignments using blat 
2. Sort the output file
3. The output file can be filtered using UCSC pslCDnaFilter

Licence
-------
Please note that commercial download and installation of the Blat and In-Silico PCR software may be licensed through Kent Informatics (http://www.kentinformatics.com).

### CODE REVIEW

**Comment by bgruening** [comment]
Thanks a lot @Yating-L!

File: tools/ucsc_blat/.shed.yml [hunk 1]
```diff
@@ -0,0 +1,10 @@
+name: ucsc_blat
+owner: yating-l
+description: Standalone blat sequence search command line tool
+homepage_url: http://genome.ucsc.edu/goldenPath/help/blatSpec.html
+type: unrestricted
+categories: 
+- Sequence Analysis
+exclude:
```
Comment (yhoogstrate) [inline tools/ucsc_blat/.shed.yml]: This is already covered by the `.gitignore` file and cna be removed.

File: tools/ucsc_blat/.shed.yml [hunk 2]
```diff
@@ -0,0 +1,10 @@
+name: ucsc_blat
+owner: yating-l
```
Comment (yhoogstrate) [inline tools/ucsc_blat/.shed.yml]: (Just as a routine check) Did you go to [Repository Action] -> [Manage repository administrators] in both tool sheds and `iuc@galaxyproject.irg`?

File: tools/ucsc_blat/blat.xml [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+            <param type="select" name="noHead" format="text" label="Suppresses .psl header (so it's just a tab-separated file)." >
+                  <option value="yes" selected="true">Yes, suppresses .psl header</option>
+                  <option value="no">No, do not suppresses .psl header</option>
+            </param>
+            <param name="mask" type="select" label="Mask out repeats" help="Alignments won't be started in masked region
+                  but may extend through it in nucleotide searches.  Masked areas
+                  are ignored entirely in protein or translated searches. Default is lower" argument="-mask">
+                  <option value="lower" selected="true">lower - mask out lower-cased sequence</option>
+                  <option value="upper">upper - mask out upper-cased sequence</option>
+                  <option value="out">out - mask according to database.out RepeatMasker .out file</option>
```
Comment (yhoogstrate) [inline tools/ucsc_blat/blat.xml]: @Yating-L sorry to ask.. Could you tell me how this `database.out` file is provided for options `out` and `file.out`?

File: tools/ucsc_blat/blat.xml [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+                  <option value="file.out">file.out - mask database according to RepeatMasker file.out</option>
+            </param>
+      </inputs>
+      <outputs>
+            <data format="psl" name="output_sorted" label="${tool.name} on ${on_string}"></data>
+      </outputs>
+    <tests>
+        <test>
+            <param name="reference_source_selector" value="history" />
+            <param name="database" value="amaVit1_Gallus/amaVit1.fa" />
```
Comment (yhoogstrate) [inline tools/ucsc_blat/blat.xml]: The size of the test is a bit too nthousiastic for this repo I suppose... Could you change this to simpler tests (e.g. to `./tools/hisat2/test-data/phiX.fa`) ?

File: tools/ucsc_blat/tool_dependencies.xml [hunk 5]
```diff
@@ -0,0 +1,27 @@
+<?xml version="1.0"?>
```
Comment (yhoogstrate) [inline tools/ucsc_blat/tool_dependencies.xml]: I have changed the requirement to a bio-conda requirement so this file can be removed. If your galaxy instance is up to date, it is able to automatically test and install blat

File: tools/ucsc_blat/blat.xml (hunk 4 above)
Comment (Yating-L) [inline tools/ucsc_blat/blat.xml]: You're right, the size of test data was too big. I picked a subset for testing, see test-data/dbia3.

File: tools/ucsc_blat/tool_dependencies.xml (hunk 5 above)
Comment (Yating-L) [inline tools/ucsc_blat/tool_dependencies.xml]: Planemo didn't pass the third test with bioconda package. However, all tests were passed with toolshed package. I am still trying to figure out why.

File: tools/ucsc_blat/blat.xml (hunk 3 above)
Comment (Yating-L) [inline tools/ucsc_blat/blat.xml]: That's a good question! I didn't consider those two options before. Now I added a place to put masking file when selecting "file.out". Option "out" means the masking file with extension .out is in the database.

File: tools/ucsc_blat/.shed.yml (hunk 2 above)
Comment (Yating-L) [inline tools/ucsc_blat/.shed.yml]: Just added iuc@galaxyproject.org as administrator in both tool sheds

File: tools/ucsc_blat/blat.xml (hunk 4 above)
Comment (Yating-L) [inline tools/ucsc_blat/blat.xml]: You're right, the size of test data was too big. I picked a subset for testing, see test-data/dbia3.

File: tools/ucsc_blat/tool_dependencies.xml (hunk 5 above)
Comment (yhoogstrate) [inline tools/ucsc_blat/tool_dependencies.xml]: Sorry for the late response. I saw the error too and I tested it manually from the command line 

`blat --arg1 --arg2 query.fa ref.2bit` <-- this worked

then I made a simlink:
`ln -s ref.2bit ref.2bit.fa`

followed by:
`blat --arg1 --arg2 query.fa ref.2bit.fa` <-- this gave the error

blat determines the file format of the database by file extension :sweat: . I have added a workaround.

File: tools/ucsc_blat/blat.xml (hunk 3 above)
Comment (yhoogstrate) [inline tools/ucsc_blat/blat.xml]: Perfect

File: tools/ucsc_blat/blat.xml (hunk 4 above)
Comment (yhoogstrate) [inline tools/ucsc_blat/blat.xml]: `du -h -d 1` still gives 28 mb. This is too large... If I find some time later this week I will try to truncate the data and I think we need to collapse/squash these commits too, to erase them from history.

**Comment by bgruening** [comment]
@Yating-L thanks a lot and sorry that it took so long at the end.

**Comment by bgruening** [comment]
@yhoogstrate thanks for the great review!

**Comment by blankenberg** [comment]
was updating an old branch of mine and noticed this:
````
Counting objects: 12251, done.
Delta compression using up to 8 threads.
Compressing objects: 100% (4062/4062), done.
Writing objects: 100% (12251/12251), 80.43 MiB | 11.08 MiB/s, done.
Total 12251 (delta 8249), reused 11998 (delta 8062)
remote: Resolving deltas: 100% (8249/8249), completed with 388 local objects.
remote: warning: GH001: Large files detected. You may want to try Git Large File Storage - https://git-lfs.github.com.
remote: warning: See http://git.io/iEPt8g for more information.
remote: warning: File tools/ucsc_blat/test-data/dbia3/dmel-all-transcript-r6.13.fasta is 97.05 MB; this is larger than GitHub's recommended maximum file size of 50.00 MB
To https://github.com/blankenberg/tools-iuc.git
```
the test file probably should have been rebased/etc out instead of just `rm`ed.  

Maybe worth adding a filesize checker to the testing proceedure?

**Pull request merged**